aaelf64-morello/aaelf64-morello.rst

```diff
@@ -865,3 +929,195 @@ veneer is used. The BX changes the execution state from A64 to C64:
+instruction sequence could be:
+
+.. code-block:: text
+
+  mrs c0, CTPIDR_EL0
+  movz x8, #:tprel_g1:local_exec_var
+  movk x8, #:tprel_g0_nc:local_exec_var
+  movz x9, #:size_g1:local_exec_var
+  movk x9, #:size_g0_nc:local_exec_var
+  add c0, c0, x8, uxtx
+  scbnds c0, c0, x9
+
+Initial Exec
+^^^^^^^^^^^^
+
+The capability to the TLS variable is derived from ``CTPIDR_EL0``. The size
+and offset of the TLS variable is stored in a GOT slot (first 8 bytes contains the
+offset and the second 8 bytes the size). This GOT slot is initialized by a
+``R_MORELLO_TPREL128`` dynamic relocation. The access must use the
+``R_MORELLO_TLSIE_ADR_GOTTPREL_PAGE20`` and ``R_MORELLO_TLSIE_ADD_LO12``
+relocations in order to allow relaxation to Local Exec. There are no other
+requirements on how this is performed or the registers used. A possible
+instruction sequence could be:
+
+.. code-block:: text
+
+  adrp c0, :gottprel:initial_exec_var
+  add c0, c0, :gottprel_lo12:initial_exec_var
+  ldp x0, x8, [c0]
+  mrs c1, CTPIDR_EL0
+  add c0, c1, x0, uxtx
+  scbnds c0, c0, x8
+
+Initial Exec to Local Exec relaxation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The linker will generate 16 bytes in the ``.rodata`` section, containing
+the offset in the static TLS block in the first 8 bytes and the size of the
+symbol in the next 8 bytes:
+
+.. code-block:: text
+
+  .section .rodata
+  _sym_data:
+    .xword tlsoffset(sym)
+    .xword sizeof(sym)
+
+.. note::
+
+  ``tlsoffset(sym)`` denotes the offset in the static TLS block of the symbol
+  ``sym``, while sizeof(sym) denotes the size of the symbol ``sym``. These are
+  not valid assebler directives.
+
+The relaxation is performed by:
+
+- changing the ``R_MORELLO_TLSIE_ADR_GOTTPREL_PAGE20`` relocation on the symbol
+  ``sym`` to a ``R_MORELLO_ADR_PREL_PG_HI20`` with the symbol ``_sym_data``
+
+- changing the ``R_MORELLO_TLSIE_ADD_LO12`` relocation on symbol the ``sym`` to
+  a ``R_AARCH64_ADD_ABS_LO12`` relocation with the symbol ``_sym_data``.
+
+General Dynamic
+^^^^^^^^^^^^^^^
+
+The instruction sequence used for the General Dynamic access model is similar to
+that of other TLSDESC implementations, with the exception that the result
+doesn't need to be added to the thread pointer. However c2 needs to contain
+the thread pointer. The instruction sequence contains an additional NOP
+instruction in order to permit the static linker to perform a relaxation to Local
+Exec or Initial Exec.
+
+The General Dynamic access sequence must be output in the following form to
+allow correct linker relaxation:
+
+.. code-block:: text
+
+  adrp c0, :tlsdesc:sym
+  ldr c1, [c0, :tlsdesc_lo12:sym]
+  add c0, c0, :tlsdesc_lo12:sym
+  nop
+  .tlsdesccall sym
+  blr c1
+
+General Dynamic to Initial Exec relaxation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The relaxed sequence is:
+
+.. code-block:: text
+
+  adrp c0, :gottprel:sym
+  add c0, [c0, :gottprel_lo12:sym]
+  ldp x0, x1, [c0]
```

Comment:
I think typo having the square-braces in the `add` instruction syntax.
Similar in the GD -> LE relaxed sequence below.